gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisherWithHiveRegistration.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.publisher;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Set;
+
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Sets;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.util.ParallelRunner;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * An extension to {@link BaseDataPublisher} for Hive registration.
+ *
+ * <p>
+ *   This publisher records the paths that need to be registered with Hive
+ *   and adds them to a property in the task state in {@link #close}.
+ * </p>
+ *
+ * @author ziliu
+ */
+@Slf4j
+public class BaseDataPublisherWithHiveRegistration extends BaseDataPublisher {
+
+  protected Set<Path> pathsToRegister = Sets.newHashSet();
+
+  public BaseDataPublisherWithHiveRegistration(State state) throws IOException {
+    super(state);
+  }
+
+  @Override
+  protected void movePath(ParallelRunner parallelRunner, Path src, Path dst, int branchId) throws IOException {
+    pathsToRegister.addAll(recordPathsToRegister(src, dst, branchId));
+    super.movePath(parallelRunner, src, dst, branchId);
+  }
+
+  @SuppressWarnings("deprecation")
+  protected Collection<Path> recordPathsToRegister(Path src, Path dst, int branchId) throws IOException {
```

Comment:
`Source` doesn't have the information of which paths the data is published to. For example, the data may be bucketed by the timestamp of each extracted record, which is not available to `Source`.